core/src/main/java/hudson/model/Executor.java

```diff
@@ -839,17 +840,37 @@ public void doStop( StaplerRequest req, StaplerResponse rsp ) throws IOException
     }
 
     /**
-     * Stops the current build.
+     * Stops the current build.<br>
+     * You can use {@link #doStopBuild(String)} instead to ensure what will be
+     * interrupted is actually what you want to interrupt.
      *
      * @since 1.489
+     * @see #doStopBuild(String)
      */
     @RequirePOST
     public HttpResponse doStop() {
+        return doStopBuild(null);
+    }
+
+    /**
+     * Stops the current build, if matching the specified external id
+     * (or no id is specified, or the current {@link Executable} is not a {@link Run}).
+     *
+     * @param runExtId
+     *      if not null, the externalizable id ({@link Run#getExternalizableId()})
+     *      of the build the user expects to interrupt
+     * @since TODO
+     */
+    @RequirePOST
+    public HttpResponse doStopBuild(@CheckForNull @QueryParameter(fixEmpty = true) String runExtId) {
         lock.writeLock().lock(); // need write lock as interrupt will change the field
         try {
             if (executable != null) {
-                getParentOf(executable).getOwnerTask().checkAbortPermission();
-                interrupt();
+                if (runExtId == null || runExtId.isEmpty()
+                        || (executable instanceof Run && runExtId.equals(((Run<?,?>) executable).getExternalizableId()))) {
+                    getParentOf(executable).getOwnerTask().checkAbortPermission();
```

Comment:
isn't it safer to keep the former behavior in case it's not a `Run` object? To avoid modifying the old behavior.

Reply:
OK, see 275b6ca. If the `executable` is not a `Run`, then it will always get interrupted, without considering the `runExtId` parameter (which covers the unluncky case of something which is not a `Run` but still has a `getExternalizableId()` method).